litr/src/main/java/com/linkedin/android/litr/io/AudioRecordMediaSource.kt

```diff
@@ -0,0 +1,179 @@
+            // Stop recording.
+            audioRecord?.stop()
+            isRecording = false
+        }
+    }
+
+    override fun getOrientationHint() = 0
+
+    override fun getTrackCount() = 1
+
+    override fun getTrackFormat(track: Int) = MediaFormat.createAudioFormat(
+            MimeType.AUDIO_RAW,
+            sampleRate,
+            channelCount
+    ).apply {
+        setInteger(MEDIA_FORMAT_PCM_KEY, AudioFormat.ENCODING_PCM_16BIT)
+    }
+
+    override fun selectTrack(track: Int) {
+        // Since we only support a single (audio) track, there is nothing to select.
+    }
+
+    override fun seekTo(position: Long, mode: Int) {
+        // We don't support seeking, since we're recording live.
+    }
+
+    override fun getSampleTrackIndex() = 0
+
+    override fun readSampleData(buffer: ByteBuffer, offset: Int): Int {
+        // If the recording has been stopped, then there are no more samples to be read.
+        if (!isRecording) {
+            return -1
+        }
+
+        // Read the next audio sample from the recorder. As per the AudioRecord documentation, we
+        // need to leave some buffer for the AudioRecord to continue to queue too. We also don't
+        // want to block waiting for more data, we'll take whatever is available.
+        val targetSize = (bufferSize / 2).coerceAtMost(buffer.capacity())
+        val readBytes = audioRecord?.read(buffer, targetSize, AudioRecord.READ_NON_BLOCKING) ?: -1
+
+        // We look at how much data has been returned, to know the duration of that sample. We
+        // therefore know when the next sample will start, which we track.
+        sampleIncrementUs = (readBytes / bytesPerUs).toLong()
+        return readBytes
+    }
+
+    override fun getSampleTime(): Long {
+        // If the recording has been stopped, then there are no more samples to be read.
+        if (!isRecording) {
+            return -1
+        }
+
+        val sampleTime = nextSampleTimeUs
+        nextSampleTimeUs += sampleIncrementUs
+        totalDurationUs += sampleIncrementUs
+        return sampleTime
+    }
+
+    override fun getSampleFlags(): Int {
+        // If the recording has been stopped, then any further samples read should report the end of
+        // the stream.
+        if (!isRecording) {
+            return MediaCodec.BUFFER_FLAG_END_OF_STREAM
+        }
+
+        return 0
+    }
+
+    override fun advance() {
+        // Nothing to advance.
+    }
+
+    override fun release() {
+        audioRecord?.stop()
+        audioRecord?.release()
+        audioRecord = null
+
+        nextSampleTimeUs = 0L
+        sampleIncrementUs = 0L
+
+        if (totalDurationUs > 0L) {
+            Log.i(TAG, "Audio Duration: $totalDurationUs")
+        }
+
+        totalDurationUs = 0L
+    }
+
+    override fun getSize() = -1L
+
+    companion object {
+        private val TAG = AudioRecordMediaSource::class.simpleName
+
+        private const val MEDIA_FORMAT_PCM_KEY = "pcm-encoding"
```

Comment:
private `val`s don't really have to be inside companion object

Reply:
Will do. Have checked a few other files and noted that you move them before the class definition (e.g. `AudioOverlayFilter`). Will copy

Reply:
Yup. Kotlin doesn't really like static/companion things. Better use `const val` in class.